core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -1594,6 +1598,128 @@ private RecordCursorFactory createSpliceJoin(
+     * </ol>
+     * <p>
+     * Detection isn't foolproof and can result in false positives. We do not attempt it unless
+     * the hint {@value SqlHints#MARKOUT_HORIZON_HINT} is present in the query.
+     * <p>
+     * These are the prerequisites for the Markout Horizon optimization to kick in:
+     * <ol>
+     *  <li><code>markout_horizon(lhs rhs)</code> hint is present
+     *  <li>LHS (master) cursor supports random access
+     *  <li>RHS (slave) cursor originates from the <code>long_sequence</code> function
+     *  <li>LHS table has a designated timestamp column
+     *  <li><code>ORDER BY</code> is a sum of two operands
+     *  <li>one operand is the designated timestamp of LHS table
+     *  <li>the other operand is a column of the RHS table
+     * </ol>
+     *
+     * @param masterAlias         alias for the master LHS of the join
+     * @param masterModel         QueryModel for the LHS of the join
+     * @param masterMetadata      Metadata for the LHS of the join
+     * @param masterCursorFactory the LHS cursor factory
+     * @param slaveModel          QueryModel for the RHS of the join
+     * @param slaveMetadata       Metadata for the RHS of the join (should be an arithmetic sequence)
+     * @param slaveCursorFactory  the RHS cursor factory (should produce an arithmetic sequence)
+     * @return MarkoutHorizonInfo if pattern is detected, null otherwise
+     */
+    private MarkoutHorizonInfo detectMarkoutHorizonPattern(
+            CharSequence masterAlias,
+            QueryModel masterModel,
+            RecordMetadata masterMetadata,
+            RecordCursorFactory masterCursorFactory,
+            QueryModel slaveModel,
+            RecordMetadata slaveMetadata,
+            RecordCursorFactory slaveCursorFactory
+    ) {
+        // Detect the markout_horizon hint, and ensure the master cursor supports random access
+        if (!SqlHints.hasMarkoutHorizonHint(masterModel, masterAlias, slaveModel.getName()) ||
+                !masterCursorFactory.recordCursorSupportsRandomAccess()
+        ) {
+            return null;
+        }
+
+        // Ensure the slave table is `long_sequence()`, or based on it
+        RecordCursorFactory rootFac = slaveCursorFactory;
+        while (!rootFac.getClass().getSimpleName().contains("LongSequence")) {
+            rootFac = rootFac.getBaseFactory();
+            if (rootFac == null) {
+                return null;
+            }
+        }
+        // Ensure the master table has a designated timestamp column
+        int designatedTimestampIndex = masterMetadata.getTimestampIndex();
+        if (designatedTimestampIndex == -1) {
+            return null;
+        }
+
+        // Ensure we have an ORDER BY clause.
+        // lastSeenOrderByModel was captured as we were descending through nested models
+        if (lastSeenOrderByModel == null) {
+            return null;
+        }
+        ExpressionNode orderByNode = lastSeenOrderByModel.getOrderBy().getQuick(0);
```

Comment:
Don't we need to check the number of ORDER BY columns?
```sql
WITH offsets AS (
    SELECT 1_000_000 * (x-1) usec_offs
    FROM long_sequence(3)
)
SELECT /*+ markout_horizon(orders offsets) */ id, order_ts + usec_offs AS ts
FROM orders CROSS JOIN offsets
ORDER BY order_ts + usec_offs, id desc;
```
seems to pass the checks while it shouldn't.